File: frontend/utils/tweet_handler.ts
import { Tweet, TweetBody } from '../typings';
import { client } from './axios';

export const fetchTweets = async () => {
    try {
      const data = await fetch('http://localhost:8082/api/v1/tweets', {
        method: 'GET',
        headers: {
         'accept': 'application/json'
        }
      })
      console.log("tweets", data)
      const tweets: Tweet[] = await data.json()
      console.log('tweets fetched succesfully;')
      return tweets
    } catch (err) {
      console.error(err) //
      return null;
  }}

export const postTweet = async (tweetBody:TweetBody) => {
    try {
      const endpoint = 'http://localhost:8082/api/v1/tweets'

      const result = await fetch(endpoint, {
        method: 'post',
        headers: {
          'Content-type': 'application/json',
        },
        body: JSON.stringify({tweetBody})
      })
      return result
    } catch (err) {
      console.error(err) //
  }}

File: backend/tweet_service/src/controllers/tweet_handler.go
package controllers

import (
	"net/http"

	"tweet_service/src/models"

	"github.com/gin-gonic/gin"
)

// @BasePath /api/v1

// Create Tweet endpoint godoc
// @Summary Create Tweet Endpoint
// @Schemes
// @Description Create Tweet
// @Tags tweet
// @Accept json
// @Produce json
// @Param models.CreateTweet body models.CreateTweet true "Create Tweett"
// @Success 200 {string} tweets
// @Router /tweets [post]
func ApiCreateTweetHandler(c *gin.Context) {

	var input models.CreateTweet

	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	db, err := models.OpenDB()
	if err != nil {
		return
	}

	handler := models.NewTweetHandler(db)

	tweets, err := handler.CreateTweet(&input)

	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "could not create tweet."})
		return
	} else {
		c.JSON(http.StatusOK, gin.H{"body": tweets})
	}
}

// @BasePath /api/v1

// Fetch Tweets endpoint godoc
// @Summary Fetch Tweets Endpoint
// @Schemes
// @Description Fetch Tweets
// @Tags tweet
// @Accept json
// @Produce json
// @Success 200 {string} body
// @Router /tweets [get]
func ApiFetchAllTweetsHandler(c *gin.Context) {

	db, err := models.OpenDB()
	if err != nil {
		return
	}

	handler := models.NewTweetHandler(db)

	tweets, err := handler.FetchAllTweets()

	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "could not fetch tweets."})
		return
	} else {
		c.JSON(http.StatusOK, tweets)
	}
}
